lib/handler/h2olog.c

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2022 Fastly, Inc, Goro Fuji, Kazuho Oku
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+#include "h2o.h"
+
+struct st_h2o_log_handler_t {
+    h2o_handler_t super;
+} h2o_log_handler;
+
+static h2o_iovec_t get_query_param(const h2o_req_t *req, const char *name, size_t name_len)
+{
+    if (req->query_at == SIZE_MAX)
+        goto NotFound;
+    const char *query = req->path.base + req->query_at;
+    const char *end = req->path.base + req->path.len;
+    while (query != end) {
+        const char *eq = strchr(query, '=');
+        if (eq == NULL)
+            goto NotFound;
+
+        if (h2o_memis(query, query - eq, name, name_len) == 0) {
+            const char *value = eq + 1;
+            const char *next_and = strchr(value, '&');
+            size_t value_len = next_and == NULL ? ((req->path.base + req->path.len) - value) : (next_and - value);
+            return h2o_iovec_init(value, value_len);
+        }
+        query = strchr(query, '&');
+        if (query == NULL)
+            goto NotFound;
+        query += 1;
+    }
+
+NotFound:
+    return h2o_iovec_init(NULL, 0);
+}
+
+static int on_req(h2o_handler_t *_self, h2o_req_t *req)
+{
+    if (req->conn->callbacks->steal_socket == NULL)
+        goto Error;
+    int req_version = req->version;
+    if (req_version != 0x100 && req_version != 0x101 && req->scheme->is_ssl)
+        goto Error;
```

Comment:
How about letting `steal_socket` determine if the socket can be provided to the handler?

FWIW, `!req->scheme->is_ssl` does not mean that the socket is non-SSL. If the request line is like `GET https://example.com/ HTTP/1.1`, then the attribute evaluates to true (because that is what the "request" says).

Therefore, we sshould be calling `h2o_socket_is_ssl` inside the `steal_socket` callback of the HTTP1 handler.